plugins/cloudservices/plugin.js

```diff
@@ -0,0 +1,127 @@
+/**
+ * @license Copyright (c) 2003-2017, CKSource - Frederico Knabben. All rights reserved.
+ * For licensing, see LICENSE.md or http://ckeditor.com/license
+ */
+
+( function() {
+	'use strict';
+
+	CKEDITOR.plugins.add( 'cloudservices', {
+		requires: 'filetools',
+		onLoad: function() {
+			var FileLoader = CKEDITOR.fileTools.fileLoader;
+
+			/**
+			 * Dedicated uploader type for [Cloud Services](https://ckeditor.com/ckeditor-cloud-services/).
+			 *
+			 * Note that this type is defined in {@link CKEDITOR.pluginDefinition#onLoad plugin.onLoad} method, thus is
+			 * guaranteed to be available in dependent plugin's {@link CKEDITOR.pluginDefinition#beforeInit beforeInit},
+			 * {@link CKEDITOR.pluginDefinition#init init} and {@link CKEDITOR.pluginDefinition#afterInit} methods.
+			 *
+			 * @since 4.8.0
+			 * @class CKEDITOR.plugins.cloudservices.cloudServicesLoader
+			 * @extends CKEDITOR.fileTools.fileLoader
+			 * @constructor
+			 * @inheritdoc
+			 * @param {CKEDITOR.editor} editor The editor instance. Used only to get language data.
+			 * @param {Blob/String} fileOrData A [blob object](https://developer.mozilla.org/en/docs/Web/API/Blob) or a data
+			 * string encoded with Base64.
+			 * @param {String} [fileName] The file name. If not set and the second parameter is a file, then its name will be used.
+			 * If not set and the second parameter is a Base64 data string, then the file name will be created based on
+			 * the {@link CKEDITOR.config#fileTools_defaultFileName} option.
+			 * @param {String} [token] A token used for [Cloud Service](https://ckeditor.com/ckeditor-cloud-services/) request. If
+			 * skipped {@link CKEDITOR.config#cloudServices_token} will be used.
+			 */
+			function CloudServicesLoader( editor, fileOrData, fileName, token ) {
+				FileLoader.call( this, editor, fileOrData, fileName );
+
+				/**
+				 * Custom [Cloud Service](https://ckeditor.com/ckeditor-cloud-services/) token.
+				 *
+				 * @property {String} custom_token
+				 * @member CKEDITOR.plugins.cloudservices.cloudServicesLoader
+				 */
+				this.customToken = token;
```

Comment:
Maybe name it just `this.token` and move here assigning custom token or default one? Then setting appropriate header in `fileUploadRequest` could be simplify to just setting it to `fileLoader.token`.

Reply:
There's a nice thing about the current implementation that it will work correctly even though config will be changed at runtime.

Officially we do not support changing config at a runtime, but it's nice to keep it whenever possible.

So the way I saw it is that `customToken` would only be set if it should really be overriden, otherwise it should be used from a single source of truth, which is config ATM.

Do you think we should change this?

Reply:
Keeping ability to change config at runtime sounds reasonable to me. Let's keep it